src/envs/screws.py

```diff
@@ -0,0 +1,501 @@
+"""Toy screw picking environment."""
+
+from typing import ClassVar, Dict, List, Optional, Sequence, Set, Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+from gym.spaces import Box
+
+from predicators.src import utils
+from predicators.src.envs import BaseEnv
+from predicators.src.settings import CFG
+from predicators.src.structs import Action, Array, GroundAtom, Image, Object, \
+    ParameterizedOption, Predicate, State, Task, Type
+
+
+class ScrewsEnv(BaseEnv):
+    """Toy screw picking env.
+
+    In this environment, the agent controls a magnetic gripper that can
+    be used to pick up screws of different colors/sizes. The challennge
+    is for the agent to learn actions that don't precisely model all the
+    irrelevant screws that might be picked up when picking a desired
+    screw.
+    """
+    _gripper_width: ClassVar[float] = 0.5
+    _receptacle_width: ClassVar[float] = 0.6
+    _receptacle_height: ClassVar[float] = 0.2
+    _screw_width: ClassVar[float] = 0.05
+    _screw_height: ClassVar[float] = 0.05
+    _magnetic_field_dist: ClassVar[float] = 0.5
+    _num_screw_clusters: ClassVar[int] = 3
+    num_screws_train: ClassVar[List[int]] = CFG.screws_num_screws_train
+    num_screws_test: ClassVar[List[int]] = CFG.screws_num_screws_test
+    # Reachable zone boundaries.
+    rz_x_lb = -5.0
+    rz_x_ub = 5.0
+    rz_y_lb = 0.0
+    rz_y_ub = 5.0
+
+    def __init__(self) -> None:
+        super().__init__()
+        # Types
+        self._screw_type = Type(
+            "screw", ["pose_x", "pose_y", "width", "height", "held"])
+        self._gripper_type = Type("gripper", ["pose_x", "pose_y", "width"])
+        self._receptacle_type = Type("receptacle",
+                                     ["pose_x", "pose_y", "width"])
+        # Predicates
+        self._ScrewPickupable = Predicate(
```

Comment:
is this the same thing as the screw being on the ground? if so, rename to `ScrewOnGround`?

Reply:
no - this means that the gripper is hovering over the screw and close enough to pick it up. I originally called it `GripperAboveScrew`, but the problem is that it's not just that, but also that the gripper is close enough to magnetize. Maybe I'll call it `GripperCanPickScrew`?

Reply:
got it, yeah `GripperCanPickScrew` sounds good